### ISSUE TITLE
Add support for std::string_view in Rcpp::as

### ISSUE THREAD

**Issue by acozzette** [RcppCore/Rcpp#1084]
I would like to add support for the std::string_view type (new in C++17) to Rcpp::as.

My motivation is that I'm working on C++ protobuf, trying to update [ServiceDescriptor::FindMethodByName()](https://github.com/protocolbuffers/protobuf/blob/2d654e8198a0eb6f9e228f964a70f15abd97512f/src/google/protobuf/descriptor.h#L1270) to take std::string_view. RProtobuf calls that function [here](https://github.com/eddelbuettel/rprotobuf/blob/868ad27b2f6028827f2f36fa531ef58287fa72db/src/RcppMacros.h#L147) (see also [here](https://github.com/eddelbuettel/rprotobuf/blob/868ad27b2f6028827f2f36fa531ef58287fa72db/src/wrapper_ServiceDescriptor.cpp#L15)), and relies on the Rcpp converter to convert to a string to pass to FindMethodByName(). I found that this stopped working once I updated that function to take std::string_view instead of const std::string&.

I will send out a pull request in just a minute, but I wanted to open this issue first as per the contributing guidelines.

**Comment by eddelbuettel** [comment]
I appreciate you filing an issue ticket to start a discussion.

Opening a pull request _concurrently_ is however the exact opposite and now we prefer.

Next up, the lowest common denominator is what many users may have. C++11 is now the bar for R 4.0.0 -- but there are many older installations still out there.  

You can use C++17 happily in a package.  We do so in [RcppSimdJson](https://github.com/eddelbuettel/rcppsimdjson).  But it is unlikely to be useful _as an Rcpp package component_.

**Comment by eddelbuettel** [comment]
Also, I don't see why we can't just deal with that inside of the [RProtoBuf[(https://github.com/eddelbuettel/rprotobuf) package?

**Comment by acozzette** [comment]
@eddelbuettel Thank you for the fast response! Sorry if I jumped the gun on the pull request, I just thought it might help to show concretely what I had in mind. I can delete it if that is the wrong direction.

I think this could likely be addressed in RProtoBuf. One way to do it would be to change this [line](https://github.com/eddelbuettel/rprotobuf/blob/868ad27b2f6028827f2f36fa531ef58287fa72db/src/RcppMacros.h#L147) to explicitly create a std::string from the converter. The only issue is that the macro is used in one other [place](https://github.com/eddelbuettel/rprotobuf/blob/868ad27b2f6028827f2f36fa531ef58287fa72db/src/wrapper_ServiceDescriptor.cpp#L14) where we need to convert to int instead of string, so I would need to split that into two separate macros. Does that approach sound OK?

**Comment by eddelbuettel** [comment]
And of course I barked before I looked at the PR ... which is indeed nice and simple.  (A test or two would not hurt though.)  

And if it is general enough / of enough general interest we may as well do it here.  Strings are, as you know, always a challenge and we had our share of issues.  But a simple `std::string_view` may work and my initial fear of "oh dear another poke with yet another stick into the whole proxy model" was unfounded :)  

Pageing @kevinushey too to take a look time permitting,
